File: app/src/main/java/com/moufee/purduemenus/ui/menu/MenuPagerAdapter.java
package com.moufee.purduemenus.ui.menu;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentStatePagerAdapter;

import com.moufee.purduemenus.menus.DiningCourtMenu;
import com.moufee.purduemenus.menus.Location;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Created by Ben on 26/09/2017.
 */

public class MenuPagerAdapter extends FragmentStatePagerAdapter {

    private Map<String, DiningCourtMenu> diningCourtMap = new HashMap<>();
    private List<Location> mLocationList = new ArrayList<>();
    private List<DiningCourtMenu> mMenusList = new ArrayList<>();
    private int mealIndex;
    private Set<String> mFavoritesSet;
    private boolean mShowFavoriteCount = true;

    public void setFavoritesSet(Set<String> favoritesSet) {
        mFavoritesSet = favoritesSet;
        notifyDataSetChanged();
    }

    public MenuPagerAdapter(FragmentManager fm) {
        super(fm);
    }

    public void setMenus(Map<String, DiningCourtMenu> diningCourtMenus) {
        this.diningCourtMap = diningCourtMenus;
        notifyDataSetChanged();
    }

    public void setLocationList(List<Location> locationList) {
        mLocationList = locationList;
        notifyDataSetChanged();
    }

    public void setMealIndex(int mealIndex) {
        this.mealIndex = mealIndex;
        notifyDataSetChanged();
    }

    public void setShowFavoriteCount(boolean showFavoriteCount) {
        mShowFavoriteCount = showFavoriteCount;
        notifyDataSetChanged();
    }

    private String getLocationForIndex(int index) {
        return mLocationList.get(index).getName();
    }

    @Override
    public Fragment getItem(int position) {
        String name = getLocationForIndex(position);
        return MenuItemListFragment.newInstance(name, mealIndex);
    }

    @Override
    public int getCount() {
        if (diningCourtMap == null || diningCourtMap.size() == 0) {
            return 0;
        }
        return mLocationList.size();
    }

    @Override
    public int getItemPosition(@NonNull Object object) {
      /*  if (object instanceof MenuItemListFragment) {
            String locationName = ((MenuItemListFragment) object).mDiningCourtName;
            for (int i = 0; i < mLocationList.size(); i++) {
                Location location = mLocationList.get(i);
                if (location.getName().equals(locationName)) {
                    return i;
                }
            }
        }*/
        //todo: better way to determine reordering of dining courts
        return POSITION_NONE;
    }

    @Override
    public CharSequence getPageTitle(int position) {
        String title = mLocationList.get(position).getName();
        if (!mShowFavoriteCount || mFavoritesSet == null || mFavoritesSet.isEmpty())
            return title;

        int numFavorites = 0;
        if (diningCourtMap.get(title).isServing(mealIndex))
            numFavorites = diningCourtMap.get(title).getMeal(mealIndex).getNumFavorites(mFavoritesSet);

        title += " (" + numFavorites + ")";
        return title;
    }
}

File: app/src/main/java/com/moufee/purduemenus/ui/menu/MenuPagerAdapter.kt
package com.moufee.purduemenus.ui.menu

import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.viewpager2.adapter.FragmentStateAdapter
import com.moufee.purduemenus.repository.data.menus.DiningCourtMeal
import com.moufee.purduemenus.repository.data.menus.MenuItem
import com.moufee.purduemenus.repository.data.menus.Station

/**
 * Created by Ben on 26/09/2017.
 */

class MenuPagerAdapter(activity: FragmentActivity) : FragmentStateAdapter(activity) {

    private var diningCourtMeals: List<DiningCourtMeal> = emptyList()
    private var favoritesSet: Set<String> = HashSet()
    private var showFavoriteCount = true

    fun setFavoritesSet(favoritesSet: Set<String>) {
        this.favoritesSet = favoritesSet
        notifyDataSetChanged()
    }

    fun setMenus(diningCourtMenus: List<DiningCourtMeal>) {
        this.diningCourtMeals = diningCourtMenus
        notifyDataSetChanged()
    }

    fun setShowFavoriteCount(showFavoriteCount: Boolean) {
        this.showFavoriteCount = showFavoriteCount
        notifyDataSetChanged()
    }

    override fun getItemId(position: Int): Long {
        return diningCourtMeals[position].diningCourtName.hashCode().toLong()
    }

    override fun containsItem(itemId: Long): Boolean {
        return diningCourtMeals.find { it.diningCourtName.hashCode().toLong() == itemId } != null
    }

    override fun createFragment(position: Int): Fragment {
        val name = diningCourtMeals[position].diningCourtName
        return MenuItemListFragment.newInstance(name)
    }

    override fun getItemCount(): Int = diningCourtMeals.size

    fun getPageTitle(position: Int): CharSequence? {

        val diningCourtName = diningCourtMeals[position].diningCourtName
        if (!showFavoriteCount || favoritesSet.isEmpty())
            return diningCourtName

        val favoriteCount = diningCourtMeals.getOrNull(position)?.stations?.countFavorites(favoritesSet) ?: 0

        return "$diningCourtName ($favoriteCount)"
    }

    private fun List<Station>.countFavorites(favoriteItemIds: Set<String>): Int {
        val favorites: MutableSet<MenuItem> = HashSet()
        forEach { station ->
            station.items.forEach {
                if (it.id in favoriteItemIds) favorites.add(it)
            }
        }
        return favorites.size
    }
}
